main.cpp

```diff
@@ -0,0 +1,241 @@
+	std::chrono::time_point<std::chrono::system_clock> end;
+	std::chrono::duration<double> elapsed;
+
+	// front insert
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->insert(0, i);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "front: insert " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	// front remove
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->remove(0);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "front: remove " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	// middle insert
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->insert(list->length() / 2, i);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "middle: insert " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	// middle remove
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->remove(list->length() / 2);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "middle: remove " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	// back insert
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->insert(-1, i);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "back: insert " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	// back remove
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->remove(-1);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "back: remove " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	for (int i = 0; i < count; i++)
+		list->insert(0, i);
+	// peek all element
+	begin = std::chrono::system_clock::now();
+	for (int i = 0; i < count; i++)
+		list->peek(i);
+	end = std::chrono::system_clock::now();
+	elapsed = end - begin;
+	cout << "peek " << count << " elements: "
+		<< elapsed.count() << " seconds\n";
+
+	cout << "-------------\n";
+
+	delete list;
+}
+
+void test_memory(int list_type) {
+	if (list_type < 1 || list_type > 2) {
+		cout << "ERROR: test_memory(): invalid list_type\n";
+		return;
+	}
+
+	List * list;
+	switch (list_type) {
+	case 1: list = new List1; break;
+	}
+
+	cout << "--------------\n";
+	cout << "test_memory(" << list_type << ")\n";
+	cout << "--------------\n";
+
+	for (int i = 0; i < 100000000; i++)
+		list->insert(0, i);
+
+	cout << "pause: ";
+	std::cin.get();
+
+	delete list;
+
+	cout << "pause: ";
+	std::cin.get();
+
+	cout << "--------------\n";
+}
+
+// ----------------------------------------------------------------------------
+
+int main() {
+	     test_functions(1);
+	//     test_functions(2);
+
+	     test_speed(1);
+	//     test_speed(2);
+
+	     test_memory(1);
+	//     test_memory(2);
+
+	return 0;  // success
+}
```

Comment:
i assume you ran this, and it works?  if so, good job using my code :-) sometimes reading and using other people's code is difficult, but it's an essential skill.